server/mergin/sync/storages/disk.py

```diff
@@ -278,22 +311,40 @@ def construct_diff(
         uploaded_file = os.path.join(self.project_dir, upload_file.location)
         diff_name = upload_file.path + "-diff-" + str(uuid.uuid4())
         changeset = os.path.join(self.project_dir, v_name, diff_name)
-        try:
-            self.flush_geodiff_logger()
-            logging.info(f"Geodiff: create changeset {changeset} from {uploaded_file}")
-            self.geodiff.create_changeset(basefile, uploaded_file, changeset)
-            # create diff metadata as it would be created by other clients
-            diff_file = File(
-                path=diff_name,
-                checksum=generate_checksum(changeset),
-                size=os.path.getsize(changeset),
-                location=os.path.join(v_name, mergin_secure_filename(diff_name)),
-            )
-            return Ok(diff_file)
-        except (GeoDiffLibError, GeoDiffLibConflictError):
-            # diff is not possible to create - file will be overwritten
-            move_to_tmp(changeset)
-            return Err(self.gediff_log.getvalue())
+        with self.geodiff_copy(basefile) as basefile_tmp, self.geodiff_copy(
+            uploaded_file
+        ) as uploaded_file_tmp:
+            try:
+                # create changeset next to uploaded file copy
+                changeset_tmp = os.path.join(
+                    uploaded_file_tmp.replace(upload_file.location, "").rstrip(
+                        os.path.sep
+                    ),
+                    v_name,
+                    diff_name,
+                )
+                self.flush_geodiff_logger()
+                logging.info(
+                    f"Geodiff: create changeset {changeset} from {uploaded_file}"
+                )
+                self.geodiff.create_changeset(
+                    basefile_tmp, uploaded_file_tmp, changeset_tmp
+                )
+                # create diff metadata as it would be created by other clients
+                diff_file = File(
+                    path=diff_name,
+                    checksum=generate_checksum(changeset_tmp),
+                    size=os.path.getsize(changeset_tmp),
+                    location=os.path.join(v_name, mergin_secure_filename(diff_name)),
+                )
+                copy_file(changeset_tmp, changeset)
+                return Ok(diff_file)
+            except (GeoDiffLibError, GeoDiffLibConflictError) as e:
+                # diff is not possible to create - file will be overwritten
+                move_to_tmp(changeset)
```

Comment:
This is performed twice in finnally.

Reply:
here we remove copy (if even created, L340) while in the finally clause we remove _tmp file